Firestore/core/src/model/mutable_document.h

```diff
@@ -0,0 +1,232 @@
+                                       ObjectValue value);
+
+  /** Creates a new document that is known to not exisr at the given version. */
+  static MutableDocument NoDocument(const DocumentKey& document_key,
+                                    const SnapshotVersion& version);
+
+  /**
+   * Creates a new document that is known to exist at the given version but
+   * whose data is not known (e.g. a document that was updated without a known
+   * base document).
+   */
+  static MutableDocument UnknownDocument(const DocumentKey& document_key,
+                                         const SnapshotVersion& version);
+
+  /**
+   * Changes the document type to indicate that it exists and that its version
+   * and data are known.
+   */
+  MutableDocument& ConvertToFoundDocument(const SnapshotVersion& version,
+                                          ObjectValue value);
+
+  /** Changes the document type to indicate that it doesn't exist at the given
+   * version. */
+  MutableDocument& ConvertToNoDocument(const SnapshotVersion& version);
+
+  /**
+   * Changes the document type to indicate that it exists at a given version but
+   * that its data is not known (e.g. a document that was updated without a
+   * known base document).
+   */
+  MutableDocument& ConvertToUnknownDocument(const SnapshotVersion& version);
+
+  MutableDocument& SetHasCommittedMutations();
+
+  MutableDocument& SetHasLocalMutations();
+
+  const DocumentKey& key() const {
+    return key_;
+  }
+
+  const SnapshotVersion& version() const {
+    return version_;
+  }
+
+  bool has_local_mutations() const {
+    return document_state_ == DocumentState::kHasLocalMutations;
+  }
+
+  bool has_committed_mutations() const {
+    return document_state_ == DocumentState::kHasCommittedMutations;
+  }
+
+  bool has_pending_writes() const {
+    return has_local_mutations() || has_committed_mutations();
+  }
+
+  const model::ObjectValue& data() const {
+    return value_;
+  }
+
+  /**
+   * Returns the value at the given path or absl::nullopt. If the path is empty,
+   * an identical copy of the FieldValue is returned.
+   *
+   * @param field_path the path to search.
+   * @return The value at the path or absl::nullopt if it doesn't exist.
+   */
+  absl::optional<FieldValue> field(const model::FieldPath& field_path) const {
+    return value_.Get(field_path);
+  }
+
+  bool is_valid_document() const {
+    return document_type_ != DocumentType ::kInvalid;
+  }
+
+  bool is_found_document() const {
+    return document_type_ == DocumentType ::kFoundDocument;
+  }
+
+  bool is_no_document() const {
+    return document_type_ == DocumentType ::kNoDocument;
+  }
+
+  bool is_unknown_document() const {
+    return document_type_ == DocumentType ::kUnknownDocument;
+  }
+
+  friend bool operator==(const MutableDocument& lhs,
+                         const MutableDocument& rhs);
+
+  friend std::ostream& operator<<(std::ostream& os, const MutableDocument& doc);
```

Comment:
Can we have a `ToString`, and implement `<<` using that? That is what `Target` does. `<<` is not very composable.

Reply:
Done. I ended up using a `stringstream` so that I can keep the operators for DocumentState and DocumentType (which might be handy later on).

Reply:
good call.